Add support for String Typecast and Array Translation

shubhamnarlawar opened the pull request:
Following are the patches in this PR - 

-> Add reading and translation support for string, float and boolean data type

-> Add support for String Typecast from `any`.

ramakotareddy commented:
Added support for string, boolean, float in Translation side.
Added support for int, float, boolean and string Typecast from Any.
Added test cases to test new functionality.

shubhamnarlawar commented:
Please review and suggest changes on this PR.

shubhamnarlawar commented:
PING.

dulvinw commented on BIR2llvmIR/BIR.h:
```diff
@@ -296,16 +297,55 @@ class MoveInsn : public NonTerminatorInsn {
+    val.intValue = intVal;
+    typeTag = TypeTag;
+  }
+  void setFloatValue(float floatVal, TypeTagEnum TypeTag) {
```
Can't we infer the typetag based on the value? For example for `int` value we can infer the `TYPE_TAG_INT`?

dulvinw commented on BIR2llvmIR/llvmFunctionGen.cpp:
```diff
@@ -175,3 +183,21 @@ VarDecl *BIRFunction::getNameVarDecl(string opName) {
   return NULL;
 }
+
+const char *BIRFunction::getTypeNameOfTypeTag(TypeTagEnum typeTag) {
```
Shouldnt we support all the subtypes of Int here?

shubhamnarlawar replied:
Does it mean Decimal and Byte data type along with int subtypes

TYPE_TAG_INT:
 TYPE_TAG_UNSIGNED8_INT:
TYPE_TAG_UNSIGNED16_INT:
TYPE_TAG_UNSIGNED32_INT:
  TYPE_TAG_SIGNED8_INT:
  TYPE_TAG_SIGNED16_INT:
  TYPE_TAG_SIGNED32_INT:
  TYPE_TAG_DECIMAL:
  TYPE_TAG_BYTE

shubhamnarlawar replied:
@ramakotareddy - Will there be a major change in translation side for supporting sub types of int?

ramakotareddy replied:
I will add int subtypes if it is minor change else will add it in next task.

shubhamnarlawar commented:
@dulvinw - Any more suggestions on this PR apart from support for sub types of int?

dulvinw commented on BIR2llvmIR/BIR.h:
```diff
@@ -296,16 +297,55 @@ class MoveInsn : public NonTerminatorInsn {
+    val.floatValue = floatVal;
+    typeTag = TypeTag;
+  }
+  void setBoolValue(bool boolVal, TypeTagEnum TypeTag) {
```
I thought they are keeping it for the conformity. Other than for int, i believe rest of the tags are self explanatory

shubhamnarlawar replied:
Yes. We are keeping it for conformity.

manuranga replied:
I also agree with Waruna, can't we have this with only one argument ? since we know the typeTag
(btw shouldn't TypeTag has to be typeTag ?)

shubhamnarlawar replied:
The process we are following is - 

Why not set the typetag directly from the BIR instead of setting it based in assumption? Using this, we can conform BIR in every function. In future, if we want to put assert messages on the received typeTag - 

`assert(typeTag != tagEnum::TAG_ENUM_CP_ENTRY_SHAPE);`

WDYT on this? I can keep only one argument also.

(The local variable is typeTag hence the parameter is named as TypeTag for readibility purpose)

  `void setBoolValue(bool boolVal, TypeTagEnum TypeTag) {`
    `val.boolValue = boolVal;`
   ` typeTag = TypeTag;`
  `}`

warunalakshitha replied:
+1 to keep the typeTag so we can add any validation if required for typetag later.

shubhamnarlawar commented on BIR2llvmIR/llvmConstantLoadInsn.cpp:
```diff
@@ -29,24 +40,45 @@ void ConstantLoadInsn::translate(LLVMModuleRef &modRef) {
     }
+    case TYPE_TAG_STRING:
+    case TYPE_TAG_CHAR_STRING: {
+      LLVMValueRef *paramTypes = new LLVMValueRef[3];
```
Sure. How about below comment? - 

/*
1. At first, string constant is created using string value and string length
2. Using string constant, a Global Variable is created.
3. Using in bound get element pointer, Global Variable is loaded into a local variable
*/

manuranga commented on BIR2llvmIR/BIRReader.cpp:
```diff
@@ -111,6 +111,24 @@ uint32_t ConstantPoolSet::getIntCp(uint32_t index) {
   return intCp->getValue();
 }
 
+// Search string from the constant pool based on index
```
```suggestion
// Search float from the constant pool based on index
```

manuranga commented on BIR2llvmIR/BIRReader.cpp:
```diff
@@ -111,6 +111,24 @@ uint32_t ConstantPoolSet::getIntCp(uint32_t index) {
+  return floatCp->getValue();
+}
+
+// Search string from the constant pool based on index
```
```suggestion
// Search boolean from the constant pool based on index
```

dulvinw commented on BIR2llvmIR/llvmConstantLoadInsn.cpp:
```diff
@@ -29,24 +40,45 @@ void ConstantLoadInsn::translate(LLVMModuleRef &modRef) {
+          *unwrap(LLVMGetGlobalContext()),
+          StringRef(stringValue.c_str(), stringValue.length()));
+      GlobalVariable *GV =
+          new GlobalVariable(*unwrap(modRef), C->getType(), false,
```
Just for my knowledge, why do we need a global variable to hold the `string` value?

shubhamnarlawar replied:
@ramakotareddy - Can you give comment on this?

ayonam commented:
That is because all string literals need to go
        into the ".rodata" section and that will happen only if we
        declare it as a GlobalVariable.  If we declare it as a local
        variable, then its initialization needs to be done character by
        character in the generated assembly code, which will add to its
        runtime.
    
        
      
      
    
    On 19/01/21 11:03, Dulvin Witharane
      wrote:
    
    
      
      @dulvinw commented on this pull request.
      
      In BIR2llvmIR/llvmConstantLoadInsn.cpp:
      >        break;
     }
+    case TYPE_TAG_STRING:
+    case TYPE_TAG_CHAR_STRING: {
+      LLVMValueRef *paramTypes = new LLVMValueRef[3];
+      string stringValue = val.strValue->data();
+      Constant *C = llvm::ConstantDataArray::getString(
+          *unwrap(LLVMGetGlobalContext()),
+          StringRef(stringValue.c_str(), stringValue.length()));
+      GlobalVariable *GV =
+          new GlobalVariable(*unwrap(modRef), C->getType(), false,

      Just for my knowledge, why do we need a global variable to hold
        the string value?
      —
        You are receiving this because you are subscribed to this
        thread.
        Reply to this email directly, view it on GitHub, or unsubscribe.
      [
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/ballerina-platform/nballerina/pull/41#pullrequestreview-570917197",
"url": "https://github.com/ballerina-platform/nballerina/pull/41#pullrequestreview-570917197",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

shubhamnarlawar commented:
Are there any further suggestions or changes need to be done on this PR? If not, can you please merge it?

shubhamnarlawar commented:
PING.

Is anything blocking for this PR to get merge in the master? I think this PR is good to go!

ruvi-d reviewed (commented):
> PING.
> 
> Is anything blocking for this PR to get merge in the master? I think this PR is good to go!

Sorry my bad. I wrote some review comments yesterday, but haven't submitted them. Let me submit them now.

shubhamnarlawar commented:
We have pushed the code for Array Translation in this PR as a seperate commit.

ruvi-d commented:
> We have pushed the code for Array Translation in this PR as a seperate commit.
> 
> Added reading and translation support for new array, store array and load array instructions.

Can you update the Readme with instructions to compile the runtime?

dulvinw commented on BIR2llvmIR/BIRReader.cpp:
```diff
@@ -262,13 +280,44 @@ ConstantLoadInsn *ReadConstLoadInsn::readNonTerminatorInsn() {
+  case TYPE_TAG_CHAR_STRING:
+  case TYPE_TAG_STRING: {
+    uint32_t valueCpIndex = readerRef.readS4be();
+    string *strVal = new string();
+    strVal[0] = readerRef.constantPool->getStringCp(valueCpIndex);
```
+1 to use RAII.

ramakotareddy commented on BIR2llvmIR/BIRReader.cpp:
```diff
@@ -262,13 +280,44 @@ ConstantLoadInsn *ReadConstLoadInsn::readNonTerminatorInsn() {
+  case TYPE_TAG_BOOLEAN: {
     uint8_t valueCpIndex = readerRef.readU1();
-    constantloadInsn->setValue(valueCpIndex);
+    constantloadInsn->setBoolValue(valueCpIndex, typeTag);
```
Fixed.

ruvi-d reviewed (changes requested):
A few comments left by me are still open

shubhamnarlawar commented:
> A few comments left by me are still open

Pushing the remaining suggested changes.

shubhamnarlawar commented:
> > We have pushed the code for Array Translation in this PR as a seperate commit.
> > Added reading and translation support for new array, store array and load array instructions.
> 
> Can you update the Readme with instructions to compile the runtime?

Sure. I will update Readme with other changes.

shubhamnarlawar commented:
> > We have pushed the code for Array Translation in this PR as a seperate commit.
> > Added reading and translation support for new array, store array and load array instructions.
> 
> Can you update the Readme with instructions to compile the runtime?

As we have updated the Makefile, a simple `make` would build the runtime also.

shubhamnarlawar commented:
Issue #56 and #57 has been created from the suggestions of this PR.

@ruvi-d - We will send remaining changes of the PR.

This PR has become very big. Is it fine if we send separate PR addressing issues #56 and #57 later?

WDYT?

ruvi-d commented:
> Issue #56 and #57 has been created from the suggestions of this PR.
> 
> @ruvi-d - We will send remaining changes of the PR.
> 
> This PR has become very big. Is it fine if we send separate PR addressing issues #56 and #57 later?
> 
> WDYT?

Makes sense @shubhamnarlawar. Let me review the commits so far and try to pass this PR through with new issues to track any requested changes that maybe required.

ruvi-d reviewed (approved):
Good to merge PR
